toolsrc/src/tls12-download.c

```diff
@@ -0,0 +1,263 @@
+        }
+
+        narrowChars -= (int)charsWritten;
+    }
+
+    if (!HeapFree(GetProcessHeap(), 0, narrowBuffer))
+    {
+        win32_abort();
+    }
+}
+
+void write_number(const HANDLE hStdOut, DWORD number)
+{
+    wchar_t buffer[11]; // 4294967295\0
+    wchar_t* cursor = buffer + 11;
+    *--cursor = L'\0';
+    if (number == 0)
+    {
+        *--cursor = L'0';
+    }
+    else
+    {
+        do
+        {
+            *--cursor = L'0' + number % 10;
+            number /= 10;
+        } while (number != 0);
+    }
+
+    write_message(hStdOut, cursor);
+}
+
+void report_api_failure(const HANDLE hStdOut, const wchar_t* api_name)
+{
+    DWORD lastError = GetLastError();
+    write_message(hStdOut, L"While calling Windows API function ");
+    write_message(hStdOut, api_name);
+    write_message(hStdOut, L"\r\n");
+    wchar_t* message;
+    if (FormatMessageW(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_FROM_HMODULE | FORMAT_MESSAGE_ALLOCATE_BUFFER,
+                       GetModuleHandleW(L"winhttp.dll"),
+                       lastError,
+                       0,
+                       (LPWSTR)&message,
+                       0,
+                       0))
+    {
+        write_message(hStdOut, message);
+        // intentionally leaks the message buffer
+    }
+    else
+    {
+        write_message(hStdOut, L"(unknown error, FormatMessageW failed)");
+    }
+
+    write_message(hStdOut, L"\r\n");
+    FlushFileBuffers(hStdOut);
+    TerminateProcess(GetCurrentProcess(), 3);
+}
+
+#ifndef NDEBUG
+int main()
+#else // ^^^ debug // !debug vvv
+int __stdcall entry()
+#endif // ^^^ !debug
+{
+    #ifdef NDEBUG
+    __security_init_cookie();
+    #endif // ^^^ release
+
+    const HANDLE stdOut = GetStdHandle(STD_OUTPUT_HANDLE);
+    if (stdOut == INVALID_HANDLE_VALUE)
+    {
+        win32_abort();
+    }
+
+    int argc;
+    wchar_t** argv = CommandLineToArgvW(GetCommandLineW(), &argc); // intentionally leaks argv
+    if (argv == 0)
+    {
+        win32_abort();
+    }
+
+    if (argc != 4)
+    {
+        write_message(stdOut, L"Usage: tls12-download.exe DOMAIN RELATIVE-PATH OUT-FILE\r\n");
+        return 1;
+    }
+
+    const wchar_t* const domain = argv[1];
+    const wchar_t* const relative_path = argv[2];
+    const wchar_t* const out_file_path = argv[3];
+    write_message(stdOut, L"Downloading https://");
+    write_message(stdOut, domain);
+    write_message(stdOut, relative_path);
+    write_message(stdOut, L" -> ");
+    write_message(stdOut, out_file_path);
+    write_message(stdOut, L"\r\n");
+
+    HANDLE outFile = CreateFileW(out_file_path, FILE_WRITE_DATA, 0, 0, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, 0);
+    if (outFile == INVALID_HANDLE_VALUE)
+    {
+        report_api_failure(stdOut, L"CreateFileW");
+    }
+
+    BOOL results = FALSE;
+    const HINTERNET session = WinHttpOpen(
+        L"tls12-download/1.0", WINHTTP_ACCESS_TYPE_DEFAULT_PROXY, WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS, 0);

```

Comment:
We probably need to do https://github.com/microsoft/vcpkg/blob/9c63f97fa1d4915a7dfc5f32d5cdb6ee8fe9bf8b/toolsrc/src/vcpkg/base/downloads.cpp#L96
```suggestion
        L"tls12-download/1.0", IsWindows8Point1OrGreater() ? WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY : WINHTTP_ACCESS_TYPE_DEFAULT_PROXY, WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS, 0);
```
for proxy support

Reply:
`IsWindows8Point1OrGreater()` is `false` the way this program is built. Certainly for the only platform on which we intend its use.